sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -124,12 +124,45 @@ case class GroupExpression(children: Seq[Expression]) extends Expression {
   override def dataType: DataType = throw new UnsupportedOperationException
 }
 
+trait TypeConstraint {
+  self: Expression =>
+
+  def isTypeMatch: Boolean
```

Comment:
i think we should just put this in Expression itself, so every expression has to define something related to type checking.